File: app/src/main/java/top/easelink/lcg/ui/main/articles/view/ArticlesFragment.java
package top.easelink.lcg.ui.main.articles.view;

import android.content.Context;
import android.os.Bundle;
import android.view.View;
import android.widget.Toast;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.annotation.StringRes;
import androidx.core.content.ContextCompat;
import androidx.lifecycle.ViewModelProviders;
import androidx.recyclerview.widget.DefaultItemAnimator;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import com.airbnb.lottie.LottieAnimationView;
import top.easelink.framework.BR;
import top.easelink.framework.base.BaseFragment;
import top.easelink.framework.customview.ScrollChildSwipeRefreshLayout;
import top.easelink.lcg.R;
import top.easelink.lcg.databinding.FragmentArticlesBinding;
import top.easelink.lcg.ui.ViewModelProviderFactory;
import top.easelink.lcg.ui.main.articles.viewmodel.ArticlesAdapter;
import top.easelink.lcg.ui.main.articles.viewmodel.ArticlesViewModel;

import javax.inject.Inject;

import static top.easelink.lcg.ui.main.articles.viewmodel.ArticlesViewModel.FETCH_INIT;

public class ArticlesFragment extends BaseFragment<FragmentArticlesBinding, ArticlesViewModel>
        implements ArticlesNavigator{

    private static final String ARG_PARAM = "param";

    @Inject
    ViewModelProviderFactory factory;
    private String mParam;

    public static ArticlesFragment newInstance(String param) {
        Bundle args = new Bundle();
        args.putString(ARG_PARAM, param);
        ArticlesFragment fragment = new ArticlesFragment();
        fragment.setArguments(args);
        return fragment;
    }

    @Override
    public int getBindingVariable() {
        return BR.viewModel;
    }

    @Override
    public int getLayoutId() {
        return R.layout.fragment_articles;
    }

    @Override
    public ArticlesViewModel getViewModel() {
        return ViewModelProviders.of(this, factory).get(ArticlesViewModel.class);
    }

    @Override
    public void handleError(Throwable t) {
        Toast.makeText(getContext(), R.string.error, Toast.LENGTH_SHORT).show();
    }

    @Override
    public void scrollToTop() {
        getViewDataBinding().backToTop.playAnimation();
        getViewDataBinding().recyclerView.smoothScrollToPosition(0);
    }

    @Override
    public void showMessage(@StringRes int resId) {
        Toast.makeText(getContext(), resId, Toast.LENGTH_SHORT).show();
    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        getViewModel().setNavigator(this);
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        setUp();
    }

    private void setUp() {
        Bundle bundle = getArguments();
        if (bundle != null) {
            mParam = getArguments().getString(ARG_PARAM);
        }
        setupRecyclerView();
        getViewModel().initUrl(mParam);
        final ScrollChildSwipeRefreshLayout swipeRefreshLayout = getViewDataBinding().refreshLayout;
        Context context = getContext();
        swipeRefreshLayout.setColorSchemeColors(
                ContextCompat.getColor(context, R.color.colorPrimary),
                ContextCompat.getColor(context, R.color.colorAccent),
                ContextCompat.getColor(context, R.color.colorPrimaryDark)
        );
        // Set the scrolling view in the custom SwipeRefreshLayout.
        swipeRefreshLayout.setScrollUpChild(getViewDataBinding().recyclerView);
        swipeRefreshLayout.setOnRefreshListener(() -> getViewModel().fetchArticles(FETCH_INIT));
    }

    private void setupRecyclerView() {
        ArticlesAdapter mArticlesAdapter = new ArticlesAdapter(getViewModel());
        LinearLayoutManager mLayoutManager = new LinearLayoutManager(getContext());
        mLayoutManager.setOrientation(RecyclerView.VERTICAL);
        RecyclerView rv = getViewDataBinding().recyclerView;
        rv.setLayoutManager(mLayoutManager);
        rv.setItemAnimator(new DefaultItemAnimator());
        rv.setAdapter(mArticlesAdapter);
        rv.addOnScrollListener(new RecyclerView.OnScrollListener() {
            @Override
            public void onScrollStateChanged(@NonNull RecyclerView recyclerView, int newState) {
                int firstVisibleItemPosition = mLayoutManager.findFirstVisibleItemPosition();
                LottieAnimationView backToTop = getViewDataBinding().backToTop;
                if (newState == RecyclerView.SCROLL_STATE_IDLE) {
                    if (firstVisibleItemPosition == 0) {
                        backToTop.setVisibility(View.GONE);
                        backToTop.pauseAnimation();
                    } else {
                        backToTop.setVisibility(View.VISIBLE);
                    }
                } else if (newState == RecyclerView.SCROLL_STATE_DRAGGING) {
                    backToTop.setVisibility(View.GONE);
                    backToTop.pauseAnimation();
                }
            }
        });
    }
}

File: app/src/main/java/top/easelink/lcg/ui/main/articles/view/ArticlesFragment.kt
package top.easelink.lcg.ui.main.articles.view

import android.os.Bundle
import android.view.View
import androidx.core.content.ContextCompat
import androidx.lifecycle.Observer
import androidx.lifecycle.ViewModelProviders
import androidx.recyclerview.widget.DefaultItemAnimator
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import kotlinx.android.synthetic.main.fragment_articles.*
import top.easelink.framework.base.BaseFragment
import top.easelink.lcg.BR
import top.easelink.lcg.LCGApp
import top.easelink.lcg.R
import top.easelink.lcg.databinding.FragmentArticlesBinding
import top.easelink.lcg.ui.main.articles.viewmodel.ArticleFetcher
import top.easelink.lcg.ui.main.articles.viewmodel.ArticlesAdapter
import top.easelink.lcg.ui.main.articles.viewmodel.ArticlesViewModel

class ArticlesFragment : BaseFragment<FragmentArticlesBinding, ArticlesViewModel>() {

    override fun getBindingVariable(): Int {
        return BR.viewModel
    }

    override fun getLayoutId(): Int {
        return R.layout.fragment_articles
    }

    override fun getViewModel(): ArticlesViewModel {
        return ViewModelProviders.of(this).get(ArticlesViewModel::class.java)
    }

    private fun scrollToTop() {
        viewDataBinding.backToTop.playAnimation()
        viewDataBinding.recyclerView.let {
            val pos = (it.layoutManager as? LinearLayoutManager)?.findLastCompletelyVisibleItemPosition()
            if (pos != null && pos > 30) {
                it.scrollToPosition(30)
                it.smoothScrollToPosition(0)

            } else {
                it.smoothScrollToPosition(0)
            }
        }
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        setUpView()
    }

    private fun setUpView() {
        setupRecyclerView()
        arguments?.getString(ARG_PARAM)?.let {
            viewModel.articles.observe(this, Observer {articleList ->
                if (articleList.isEmpty() && viewModel.isLoading.value == true) {
                    viewDataBinding.recyclerView.visibility = View.GONE
                } else {
                    viewDataBinding.recyclerView.visibility = View.VISIBLE
                    (viewDataBinding.recyclerView.adapter as? ArticlesAdapter)?.apply {
                        if (itemCount > 1) {
                            appendItems(articleList)
                        } else {
                            clearItems()
                            addItems(articleList)
                        }
                    }
                }
            })
            viewModel.initUrl(it)
        }
        back_to_top.setOnClickListener {
            scrollToTop()
        }
        viewDataBinding.refreshLayout.run {
            val context = context?:LCGApp.getContext()
            setColorSchemeColors(
                ContextCompat.getColor(context, R.color.colorPrimary),
                ContextCompat.getColor(context, R.color.colorAccent),
                ContextCompat.getColor(context, R.color.colorPrimaryDark)
            )
            // Set the scrolling view in the custom SwipeRefreshLayout.
            setScrollUpChild(viewDataBinding.recyclerView)
            setOnRefreshListener {
                viewModel.fetchArticles(ArticleFetcher.FetchType.FETCH_INIT)
            }
        }
    }

    private fun setupRecyclerView() {
        viewDataBinding?.recyclerView?.apply {
            val mLayoutManager = LinearLayoutManager(context).also {
                it.orientation = RecyclerView.VERTICAL
            }
            layoutManager = mLayoutManager
            itemAnimator = DefaultItemAnimator()
            adapter = ArticlesAdapter(viewModel).also {
                it.setFragmentManager(fragmentManager?:baseActivity.supportFragmentManager)
            }
            addOnScrollListener(object : RecyclerView.OnScrollListener() {
                override fun onScrollStateChanged(
                    recyclerView: RecyclerView,
                    newState: Int
                ) {
                    viewDataBinding.backToTop.let {
                        if (newState == RecyclerView.SCROLL_STATE_IDLE) {
                            if (mLayoutManager.findFirstVisibleItemPosition() <= 1) {
                                it.visibility = View.GONE
                                it.pauseAnimation()
                            } else {
                                it.visibility = View.VISIBLE
                            }
                        } else if (newState == RecyclerView.SCROLL_STATE_DRAGGING) {
                            it.visibility = View.GONE
                            it.pauseAnimation()
                        }
                    }

                }
            })
        }

    }

    companion object {
        private const val ARG_PARAM = "param"
        @JvmStatic
        fun newInstance(param: String): ArticlesFragment {
            return ArticlesFragment().apply {
                arguments = Bundle().also {
                    it.putString(ARG_PARAM, param)
                }
            }
        }
    }
}